sdk/dotnet/Pulumi/Config.cs

```diff
@@ -0,0 +1,361 @@
+// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Config is a bag of related configuration state.  Each bag contains any number of
+    /// configuration variables, indexed by simple keys, and each has a name that uniquely
+    /// identifies it; two bags with different names do not share values for variables that
+    /// otherwise share the same key.  For example, a bag whose name is <c>pulumi:foo</c>, with keys <c>a</c>,
+    /// <c>b</c>, and <c>c</c>, is entirely separate from a bag whose name is <c>pulumi:bar</c> with
+    /// the same simple key names.  Each key has a fully qualified names, such as
+    /// <c>pulumi:foo:a</c>, ..., and <c>pulumi:bar:a</c>, respectively.
+    /// </summary>
+    public partial class Config

```

Comment:
A minor thing that is likely controversial that I don't feel strongly about either way but wanted to at least ask: Should we make it `sealed` by default for now to make it easier to make changes without breaking folks in the future?

Reply:
yes.  i'd like to seal most of our stuff to keep things under our control.

Reply:
done.